### PR TITLE
xl: Make locking more granular for PutObjectPart requests

### LINKED ISSUE #2082: Investigate block profile results for concurrent uploads of objects
## Setup
- Ran block profiling on #2066 

[2066-blockprof.pdf](https://github.com/minio/minio/files/345399/2066-blockprof.pdf)

### PR DESCRIPTION
**Pull request by krisis**
Closes #2082

### CODE REVIEW

File: xl-v1-healing.go [hunk 1]
```diff
@@ -47,12 +47,12 @@ func listObjectVersions(partsMetadata []xlMetaV1, errs []error) (versions []int6
 
 // Reads all `xl.json` metadata as a xlMetaV1 slice.
 // Returns error slice indicating the failed metadata reads.
-func (xl xlObjects) readAllXLMetadata(bucket, object string) ([]xlMetaV1, []error) {
-	errs := make([]error, len(xl.storageDisks))
-	metadataArray := make([]xlMetaV1, len(xl.storageDisks))
+func (xl xlObjects) readAllXLMetadata(disks []StorageAPI, bucket, object string) ([]xlMetaV1, []error) {
```
Comment (harshavardhana) [inline xl-v1-healing.go]: Remove these as being part of `(xl xlObjects)` - make them pure functions.

File: xl-v1-metadata.go [hunk 2]
```diff
@@ -272,10 +272,10 @@ func writeXLMetadata(disk StorageAPI, bucket, prefix string, xlMeta xlMetaV1) er
 }
 
 // deleteAllXLMetadata - deletes all partially written `xl.json` depending on errs.
-func (xl xlObjects) deleteAllXLMetadata(bucket, prefix string, errs []error) {
+func (xl xlObjects) deleteAllXLMetadata(disks []StorageAPI, bucket, prefix string, errs []error) {
```
Comment (harshavardhana) [inline xl-v1-metadata.go]: Same as above

File: xl-v1-metadata.go [hunk 3]
```diff
@@ -293,12 +293,12 @@ func (xl xlObjects) deleteAllXLMetadata(bucket, prefix string, errs []error) {
 }
 
 // writeUniqueXLMetadata - writes unique `xl.json` content for each disk in order.
-func (xl xlObjects) writeUniqueXLMetadata(bucket, prefix string, xlMetas []xlMetaV1) error {
+func (xl xlObjects) writeUniqueXLMetadata(disks []StorageAPI, bucket, prefix string, xlMetas []xlMetaV1) error {
```
Comment (harshavardhana) [inline xl-v1-metadata.go]: Same as above.

**Comment by krishnasrinivas** [comment]
reviewing this PR

File: xl-v1-multipart.go [hunk 4]
```diff
@@ -443,23 +435,51 @@ func (xl xlObjects) PutObjectPart(bucket, object, uploadID string, partID int, s
...
+				updatedEInfos[index].Checksum[j] = eInfo.Checksum[j]
+			}
+		}
+	}
+
+	// Pick one from the first valid metadata.
+	xlMeta := pickValidXLMeta(partsMetadata)
+
+	// Re-read higherVersio
+	_, higherVersion, err = xl.listOnlineDisks(partsMetadata, errs)
```
Comment (krishnasrinivas) [inline xl-v1-multipart.go]: here `onlineDisks` retval is being ignored

File: xl-v1-multipart.go [hunk 5]
```diff
@@ -442,25 +440,44 @@ func (xl xlObjects) PutObjectPart(bucket, object, uploadID string, partID int, s
...
+				checksumSet[cksum.Name] = cksum
+			}
+			// Form the checksumInfo to be committed in xl.json
+			// from the map.
+			var finalChecksums []checkSumInfo
+			for _, cksum := range checksumSet {
+				finalChecksums = append(finalChecksums, cksum)
+			}
+			updatedEInfos[index] = eInfo
+			updatedEInfos[index].Checksum = finalChecksums
```
Comment (krishnasrinivas) [inline xl-v1-multipart.go:472]: there is one corner case race condition:
1) upload part.1
2) upload part.2 and part.1 in parallel

there is a chance that part.1's old checksum is retained.

Comment (krishnasrinivas) [inline xl-v1-multipart.go:472]: erasureCreateFile should return []checkSumInfo instead of []erasureInfo

**Pull request merged**